### ISSUE TITLE
chezmoi externals of `git` type

### ISSUE THREAD

**Issue by felipecrs** [twpayne/chezmoi#1421]
## Is your feature request related to a problem? Please describe.

Not a big problem, but I'm recently in love with chezmoi's external feature, and would like to extend it even more. The concern I have about using archives instead of direct git repositories is that they would take a lot longer to update than when using just `git pull`, as git would only download the changes.

Another thing is that with `git` we don't need to know the name of branch, if the goal is to clone the default. If Oh-My-ZSH changes their default branch from master to main, then probably the current archive url using master would stop working.

## Describe the solution you'd like

The `git` type for `externals`, so that:

```TOML
[".oh-my-zsh"]
    # uses git to download
    type = "git"
    
    # url could also be whatever the user prefer, such as ssh://
    url = "https://github.com/ohmyzsh/ohmyzsh"
    
    ## some options could be
    
    # the branch should default to the default branch
    branch = "master"

    # allows faster first clone (especially useful for ephemeral environments)
{{ if .is_devcontainer -}}
    depth = 1
{{ end -}}

    # do the needful for cleaning up the git repo if it exists (`git clean -fdx && git reset --hard origin/$branch` for example)
    clean = "force" # options could be: true (but not forcefully -- would stop or whatever in case it cannot cleanup), false (don't do anything for trying to cleanup)

    # some way to set if we want chezmoi to always try to pull changes for this repo, or let it off for relying on --refresh-externals
    pull = "always"

    ## git clone speeds can be further optimized with some options like so

    # --single-branch
    singleBranch = true # but is already implied by git when using depth

    # --no-tags
    tags = false

    # clones only subdirectory (through git clone --filter and so: https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone/)
    filterDirs = "important-directory"
    
    # optimize for ephemeral envs(like chezmoi init --one-shot)
{{ if .is_devcontainer -}}
    oneShot = true # same as depth = 1, tags = false, singleBranch = true
{{ end -}}
```

## Describe alternatives you've considered

I'm using `antigen` for that, but it's only applicable for ZSH modules (I guess). I could easily drop the dependency on `antigen` in my dotfiles if I had such feature, which is especially meaningful as it is a step that is always run even in ephemeral environments (Codespaces, Remote Containers). It would prevent me from having to install Antigen, which mean less one call to an external server.

**Comment by twpayne** [comment]
Thanks for the proposal! This is also [something that I have considered](https://github.com/twpayne/chezmoi/blob/c03209fd11d64d59ef8e8b6436db544b12368ef4/internal/chezmoi/sourcestate.go#L4).

As you point out, there is a lot more to manage with a git external compared to a single file or archive external. It's not just the multitude of options, but also the state to be persisted on disk is no longer a single file that can be atomically replaced, but instead a collection of files that might be individually removed or corrupted.

In the end, I didn't implement git support in the first version of chezmoi externals because chemzoi's requirements of "give me the contents of this repo at this commit" are exactly answered by requesting an archive.

> The concern I have about using archives instead of direct git repositories is that they would take a lot longer to update than when using just git pull, as git would only download the changes.

I don't think this is the case. `git pull` for sure in theory only pulls the changes, but in practice there's a complex negotiation that goes on between server and client to determine which changes actually need to be sent to the client. Furthermore, the server then has to pack the changes into some sort of minimal format, and the client then needs to unpack those changes locally, and eventually generate the desired state of the repo. In contrast, requesting an archive over HTTP defers all the processing of the git repo to the server which can then deliver the final state in a widely-accepted format. For popular repos, I'm sure that GitHub caches the `master.tar.gz` archives.

> Another thing is that with `git` we don't need to know the name of branch, if the goal is to clone the default. If Oh-My-ZSH changes their default branch from master to main, then probably the current archive url using master would stop working.

Repo default branch name changes are a relatively rare event, and anyone who cares about them is certainly following the repo closely enough to make the updates to their scripts/config files when the default branch name changes.

tl;dr This is a great proposal, the cost/benefit of implementing it need to be demonstrated, keeping this issue open.

**Comment by felipecrs** [comment]
![image](https://user-images.githubusercontent.com/29582865/134429260-743051c5-6f3b-48c5-ad21-8245b78574bd.png)

The point where I want to highlight the gain is where there would not be changes to be pulled.

**Comment by twpayne** [comment]
> Also, downloading the whole archive is directly dependent on the bandwidth speed. I have a good internet here, but what if not?

For sure. That said, chezmoiexternals caches the external in `~/.cache/chezmoi/externals` and only re-downloads when you pass the `-R` flag or change the external's URL. If you want more control over the exact version you're downloading, then that's embedded in the archive URL.

**Comment by felipecrs** [comment]
chezmoi's cache does not help for assuring the external matches the state that I want: the latest.

**Comment by twpayne** [comment]
The reason that the cache is there is that chezmoi needs to evaluate the target state for many commands (e.g. `chezmoi status` and `chezmoi diff`), and re-downloading an external each time is very slow.

I wonder if rate-limited automatic refresh would be helpful, e.g. only check maximum once per day. If you want to force a refresh, you can always use the `-R` flag.

**Comment by felipecrs** [comment]
Ok, I started to convert all my ZSH plugins to this new approach. It can be seemed at: https://github.com/felipecrs/dotfiles/blob/b2ed503421b2aec905c9635c6e274872d5aaea3e/home/.chezmoiexternal.yaml#L1-L6

There is an additional problem which I caught:


```console
❯ chezmoi apply
.oh-my-zsh/cache/completions has changed since chezmoi last wrote it [overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/cache/grep-alias has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/cache/kubectl_completion has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/cache/zcomp-FELIPE-MSI has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/cache/zcomp-FELIPE-MSI.zwc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/custom/themes/powerlevel10k/gitstatus/gitstatus.plugin.zsh.zwc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/custom/themes/powerlevel10k/gitstatus/install.zwc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/custom/themes/powerlevel10k/internal/configure.zsh.zwc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/custom/themes/powerlevel10k/internal/icons.zsh.zwc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/custom/themes/powerlevel10k/internal/p10k.zsh.zwc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/custom/themes/powerlevel10k/internal/parser.zsh.zwc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/custom/themes/powerlevel10k/internal/worker.zsh.zwc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/custom/themes/powerlevel10k/powerlevel10k.zsh-theme.zwc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
.oh-my-zsh/custom/themes/powerlevel10k/powerlevel9k.zsh-theme.zwc has changed since chezmoi last wrote it [diff,overwrite,all-overwrite,skip,quit]? o
```

All these files get generated automatically, so they are not present in their original archives. But `exact = true` would still be desired for handling other files (not automatically generated).

These files are of course present in the .gitignore from their original repo.

The only clean solution to this problem that I can think of is by using `git` for handling it as well.

**Comment by felipecrs** [comment]
> The reason that the cache is there is that chezmoi needs to evaluate the target state for many commands (e.g. `chezmoi status` and `chezmoi diff`), and re-downloading an external each time is very slow.

That's absolutely understood.

**Comment by felipecrs** [comment]
Me and @twpayne discussed a bit about this feature request. Now, I think I've found a way to contemplate both issues raised:

1. For https://github.com/twpayne/chezmoi/issues/1421#issuecomment-926015535:

  chezmoi can parse the `.gitignore` files in the archive and ignore them too, as if they were part of `.chezmoiignore`.
  
2. For https://github.com/twpayne/chezmoi/issues/1421#issuecomment-950400841:

  chezmoi can similarly read the `.gitmodules` files and retrieve the list of archives on them and where to place, recursively, as if they were added to `.chezmoiexternals.yaml` by the user.
  
Both cases are situations specific to git archives and not archives in general. To decide whether 1 and 2 should be performed, I suggest we add an additional type of chezmoi external: `git-archive`.

It would behave the same as current's `archive` type, with the difference that it would do 1 and 2. Optionally, a simpler syntax in the URL could be accepted too: if it's `https://github.com/a/repo`, chezmoi could automatically infer the final archive URL. Despite this is 100% optional, the logic for it would need to be implemented anyway for achieving point 2 (which we would need to convert an entry from `.gitmodules` to a valid archive URL).

PS: all this without adding `git` as an external type, so that we would not need to change chezmoi's internal too much.

**Comment by felipecrs** [comment]
One thing to note is that, on point number 1, the `.gitignore`s found in the archive should be treated as ignore rules for chezmoi **only** during the expansion of the archive and cleaning up of the repo (`exact: true`), but not after.

Otherwise, someone adding a file manually, such as a completion, to a folder like `~/.oh-my-zsh/custom/plugins/my-completions` would end up  being ignored too, as `~/.oh-my-zsh/.gitignore` contains `custom/`.

**Comment by twpayne** [comment]
We also need to be careful about paths in `.gitignore`. For example, if the `.gitignore` file contains `foo` then it should only apply to the imported archive, not `foo` globally.

Similarly, we also need to be careful about relative paths. For example, if `.gitignore` contains `../*` then this should have no effect.

**Comment by felipecrs** [comment]
> For example, if the .gitignore file contains foo then it should only apply to the imported archive, not foo globally.

I think this is already how chezmoi works, if a chezmoiignore is inside of a subfolder (and how git with gitignore too).

> Similarly, we also need to be careful about relative paths. For example, if .gitignore contains ../* then this should have no effect.

Good point, I wonder if git supports such a case.

**Comment by twpayne** [comment]
For what it's worth, I'll probably use https://pkg.go.dev/github.com/go-git/go-git/plumbing/format/gitignore to implement this.

**Comment by twpayne** [comment]
Created #1623 to track `git-archive`.

**Comment by twpayne** [comment]
On parsing `.gitmodules` files, as @bardo points out, each `.gitmodules` file can contain further `.gitmodules` files, leading to unbounded recursion. This is much more easily solved by requiring the user to add `.chezmoiexternal` entries for each submodule.

With #1709 nearing completion, and the complexity of implementing a reliable `git` external, I'm going to close this.
